flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/JoinToMultiJoinRule.java

```diff
@@ -0,0 +1,683 @@
+                } else {
+                    for (int joinKey : joinKeys) {
+                        if (inputNullGenFieldList.get(joinKey + beginIndex)) {
+                            return false;
+                        }
+                    }
+                }
+            } else if (joinType == JoinRelType.RIGHT) {
+                if (isLeft) {
+                    return false;
+                } else {
+                    for (int joinKey : joinKeys) {
+                        if (inputNullGenFieldList.get(joinKey + beginIndex)) {
+                            return false;
+                        }
+                    }
+                }
+            } else if (joinType == JoinRelType.INNER) {
+                for (int joinKey : joinKeys) {
+                    if (inputNullGenFieldList.get(joinKey + beginIndex)) {
+                        return false;
+                    }
+                }
+            } else {
+                return false;
+            }
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Shifts a filter originating from the right child of the LogicalJoin to the right, to reflect
+     * the filter now being applied on the resulting MultiJoin.
+     *
+     * @param joinRel the original LogicalJoin
+     * @param left the left child of the LogicalJoin
+     * @param right the right child of the LogicalJoin
+     * @param rightFilter the filter originating from the right child
+     * @return the adjusted right filter
+     */
+    private RexNode shiftRightFilter(
+            Join joinRel, RelNode left, MultiJoin right, RexNode rightFilter) {
+        if (rightFilter == null) {
+            return null;
+        }
+
+        int nFieldsOnLeft = left.getRowType().getFieldList().size();
+        int nFieldsOnRight = right.getRowType().getFieldList().size();
+        int[] adjustments = new int[nFieldsOnRight];
+        for (int i = 0; i < nFieldsOnRight; i++) {
+            adjustments[i] = nFieldsOnLeft;
+        }
+        rightFilter =
+                rightFilter.accept(
+                        new RelOptUtil.RexInputConverter(
+                                joinRel.getCluster().getRexBuilder(),
+                                right.getRowType().getFieldList(),
+                                joinRel.getRowType().getFieldList(),
+                                adjustments));
+        return rightFilter;
+    }
+
+    /**
+     * Adds on to the existing join condition reference counts the references from the new join
+     * condition.
+     *
+     * @param multiJoinInputs inputs into the new MultiJoin
+     * @param nTotalFields total number of fields in the MultiJoin
+     * @param joinCondition the new join condition
+     * @param origJoinFieldRefCounts existing join condition reference counts
+     * @return Map containing the new join condition
+     */
+    private com.google.common.collect.ImmutableMap<Integer, ImmutableIntList>
```

Comment:
use the shaded guava. Or avoid it altogether and use a regular map that is guarded with `Collections.immutableMap()`. We don't use Guava directly in Flink.

Reply:
Using now the native map and doing a ImmutableMap.copy when doing new MultiJoin since the calcite class expects it